viewing_party/party.py

```diff
@@ -0,0 +1,253 @@
+from pprint import pprint
+
+#  Wave 1 **********************************************
+
+# 1
+
+def create_movie(title, genre, rating):
+    
+    movie = {}
+
+    if title is None or genre is None or rating is None:
+        return None
+
+    if title:
+        movie["title"] = title
+
+    if genre:
+        movie["genre"] = genre
+
+    if rating:
+        movie["rating"] = rating
+
+    return movie
+
+
+# 2
+
+def add_to_watched(user_data, movie):
+    user_data["watched"].append(movie)
+
+    return user_data
+
+
+# 3
+
+def add_to_watchlist(user_data, movie):
+    user_data["watchlist"].append(movie)
+
+    return user_data
+
+
+# 4
+
+def watch_movie(user_data, title):
+    watched = user_data["watched"]
+    watch_list = user_data["watchlist"]
+
+    for index, movie_info in enumerate(watch_list):
+        if title == movie_info["title"]:
+            watched.append(movie_info)
+            user_data["watched"] = watched
+            del watch_list[index]
+    return user_data
+
+
+# Wave 2 **********************************************
+
+# 1
+
+def get_watched_avg_rating(user_data):
+    watched = user_data["watched"]
+    ratings = []
+    for movie_info in (watched):
+        ratings.append(movie_info["rating"])
+
+    if len(watched) == 0:
+        return 0.0
+    else:
+        avg_rating = sum(ratings)/len(ratings)
+        return avg_rating
+
+
+# 2
+
+def get_most_watched_genre(user_data):
+    watched = user_data["watched"]
+    genres = []
+    for movie_info in watched:
+        genres.append(movie_info["genre"])
+
+    fav_genres = {}
+    if len(watched) == 0:
+        return None 
+    else:
+        for genre in genres:
+            if genre not in fav_genres:
+                fav_genres[genre] = 1
+            else:
+                fav_genres[genre] += 1
+    
+    favorite_genre = max(fav_genres, key=fav_genres.get)
+    
+    return favorite_genre
+
+
+# Wave 3 **********************************************
+
+# 1
+
+def get_unique_watched(user_data):
+    unique_watched_list = []
+    for dicts in user_data["watched"]:
+        unique_watched_list.append(dicts)
+    
+    friends_watched_list = []
+    for friends_watched in user_data["friends"]:
+        for friends_dicts in friends_watched["watched"]:
+            friends_watched_list.append(friends_dicts)
+
+    unique_watched_set = set()
+    friends_watched_set = set()
+
+    if len(user_data["watched"]) == 0:
+        return []
+    else:
+        for movie_dict in unique_watched_list:
+            for movie in movie_dict.values():
```

Comment:
Here, you're implicitly relying on the coincidence that the movie dictionaries in this wave only have a title key. The overall operation of this approach (packing values (in general of unknown type) in a set, they rehydrating them back to dictionaries at the end isn't extensible. Even though this wave only has title in the movie dictionary, notice that previous and subsequent waves had other fields as well. The tests restrict the keys in present in the dictionaries to those which are useful to focus on for a particular wave, but we should still write our code to work with movies without assuming that is the _only_ data in the movie.

But what we could do here is use sets to build unique lookup sets which we could use while iterating over the full movie records, adding those that meet our restrictions (for uniqueness, etc) to the output list. Then we don't find ourselves in a situation where we need to recreate a movie dictionary without generally knowing what might be in it, and without certainty that all the values are immutable.

But even in this case, rather than using `.values()` to iterate our way to the title, we should simply access it directly as:
```python
                unique_watched_set.add(movie_dict["title"])
```